DIPs/1NNN-AP.md

```diff
@@ -0,0 +1,257 @@
+
+A suggestion for a similar feature in Rust, but with a different approach: https://internals.rust-lang.org/t/enum-path-inference-with-variant-syntax/16851
+
+## Description
+This DIP proposes that ETI in D should use the syntax `$enumMember`. This syntax was chosen because the dot operator without an enum name as used in other languages, i.e., `.enumMember`, would have the same syntax as D's [module scope operator](https://dlang.org/spec/module.html#module_scope_operators). Alternatives are possible,
+but the proposed syntax represents the best compromise between convenience and compatibility.
+
+Following are the circumstances in which ETI will be permitted.
+
+### 1. Initializers and assignments
+ETI is allowed when initializing or assigning to a variable that has a known enum type.
+```d
+enum A{ a,b,c,d }
+
+struct S{ A one, two; }
+
+void main(){
+    A    myA1 = $b;      // myA1 = A.b
+    A    myA2 = $b | $c; // myA2 = A.c
+    auto myA3 = $b;      // error, we don't know the type of "$b"!
+
+    S myS;
+    myS.one = $c; // myB.one = A.c
+    myS.two = $d; // myB.two = A.d
+}
+```
+
+### 2. Return statements
+ETI is allowed in return statements when a function has an explicit enum return type.
+```d
+enum A{ a,b,c,d }
+
+A myFn(){
+    return $c; //returns A.c
+}
+
+auto myBrokenFn(){
+    return $c; // error, we don't know the type of "$c"!
+}
+```
+
+### 3. Argument lists
+ETI is allowed in the argument lists of function calls and template instantiations when they can bind to explicitly typed enum parameters.
+```d
+enum A{ a,b,c,d }
+
+struct S{ A one, two; }
+
+void myFn(A param){}
+void myDefaultFn(A param=$d){}
+void myTempFn(T)(T param){}
+
+void main(){
+    S    myS1 = {one:$a, two:$b};
+    auto myS2 = S($a, $b);
+
+    myFn($a);
+    myFn($a + $a); // passes A.b
+
+    myDefaultFn();
+    myDefaultFn($c);
+
+    myTempFn!A($a);
+    myTempFn($a); // error, can't infer a type to instantiate the template with from "$a"
+}
+```
+
+### 4. Case statements
+ETI is allowed in the case statements of a switch.
+```d
+import std.stdio: writeln;
+
+enum WordLetterOfTheDay{ a,b,c,d/*...*/ }
+
+void main(){
+    auto letterToday = WordLetterOfTheDay.b;
+    
+    import std.stdio;
+    switch(letterToday){
+        case $a:
+            writeln("Apple");
+            break;
+        case $b:
+            writeln("Bicycle");
+            break;
+        case $c:
+            writeln("Caterpillar");
+            break;
+        case $d:
+            writeln("Didgeridoo");
+            break;
+        /*...*/
+    }
+}
+```
+
+### 5. Array literals
+A) ETI is allowed in array literals in initializations and assignments as determined by item 1 (Initializations and assignments) above.
+
+B) ETI is also allowed in array literals for which an explicit enum type can be inferred.
+```d
+enum A{ a,b,c,d }
+
+// (A)
+A[4] x = [$a, $b, $c, $d];
+// (B)
+auto y = [A.a, $b, $c, $d]; // typeof(y) = A[]
+auto z = [A.c, 64, $b, $b]; // typeof(z) = int[]
```

Comment:
That B part opens up a bigger can of worms than you might think. What happens here?

```d
enum A{ a,b,c,d }
enum B{ d,c,b,a }

enum C{ e,f,g,h }
enum D:C{ i=C.f, j=C.g }

// int[] or ambiquous?
auto a = [A.c, 64, $b];
// int[][] or ambiquous?
auto b = [[A.a, A.b, $c], [B.a B.b, $c]];
// C[] or member h not found?
auto c = [D.i, $h];
// C[] or member j not found?
auto d = [C.e, $j];
```

Reply:
> ```d
> enum A{ a,b,c,d }
> enum B{ d,c,b,a }
> // int[] or ambiquous?
> auto a = [A.c, 64, $b];
> // int[][] or ambiquous?
> auto b = [[A.a, A.b, $c], [B.a B.b, $c]];

Ambiguity may well be heavily defined by the implementation of ETI, which is not my area unfortunately.
You can already run this code if you remove the ETI:
```d
void main(){
    auto a = [A.c, 64, A.b];
    auto b = [[A.a, A.b, A.c], [B.a, B.b, B.c]];
}
```
`a` is an int[] and `b`'s initialiser causes a compile error:
> Error: incompatible types for `([a, b, c]) : ([a, b, c])`: `A[]` and `B[]`

Would ETI's implementation turn this into an `int[][]`? I have no clue.

> ```d
> enum C{ e,f,g,h }
> enum D:C{ i=C.f, j=C.g }
> // C[] or member h not found?
> auto c = [D.i, $h];
> // C[] or member j not found?
> auto d = [C.e, $j];
> ```
I cannot see how `c` or `d`'s initialiser there could possibly compile to `C[]` **or** `D[]`, you can't use two different enum types in the same dynamic array.

Reply:
For case a, you missed the point. `a` has to be `int[]`, because the literal contains `64`. The point is, $b could be inferred as either being unambiguously `A.b`, or being ambiguous between `A.b` and `B.b`, since `B` also implicitly converts to `int`.

For case b you are correct. Arrays do not convert to their parent types even if their elements do.

> cannot see how c or d's initialiser there could possibly compile to C[] or D[], you can't use two different enum types in the same dynamic array.

Because D has a base type of C. This compiles:
```D
enum C{ e,f,g,h }
enum D:C{ i=C.f, j=C.g }
auto c = [D.i, C.h];
auto d = [C.e, D.j];
```

Reply:
> For case a, you missed the point. `a` has to be `int[]`, because the literal contains `64`. The point is, $b could be inferred as either being unambiguously `A.b`, or being ambiguous between `A.b` and `B.b`, since `B` also implicitly converts to `int`.

Why would it infer `B.b` when `A` has been used in the literal but `B` hasn't? That's not how D's inference usually works at all.

> For case b you are correct. Arrays do not convert to their parent types even if their elements do.

What do you mean I'm correct? I didn't give you a concrete answer, because your question has no answer yet.

>This compiles:
> ```d
> enum C{ e,f,g,h }
> enum D:C{ i=C.f, j=C.g }
> auto c = [D.i, C.h];
> auto d = [C.e, D.j];
> ```
To `int[]`s, because the input is nonsense. I highly doubt that would even work, but again you will have to see into the future to learn how the implementation will affect atypical use-cases like this. As the DIP author, I simply don't see the positive value in demanding a specific implementation case for the example you gave. It's too ambiguous for anyone to agree how it should work.

Reply:
> Why would it infer B.b when A has been used in the literal but B hasn't? That's not how D's inference usually works at all.

I don't think D sets any precedence. D usually infers types of individual elements without looking at other types, and only at the end decides a common element type for the array. The type of elements with ETI can only be decided after types other elements have been figured out, which is unlike anything in the current language. That's why it's not clear what would happen here.

> What do you mean I'm correct? I didn't give you a concrete answer, because your question has no answer yet.

I meant that since `[[A.a, A.b, A.c], [B.a, B.b, B.c]]` does not work (i thought it would), there is no reason why ETI should change that.

> I highly doubt that would even work, but again you will have to see into the future to learn how the implementation will affect atypical use-cases like this. As the DIP author, I simply don't see the positive value in demanding a specific implementation case for the example you gave. It's too ambiguous for anyone to agree how it should work.

It's exactly in atypical cases like this where language design errors will usually manifest. You don't know whether your idea has contradictions before you explore the more technical cases and nail the semantics down. In fact, that's why the DIPs are reviewed 3+ times in the first place.

Reply:
The bottom line is that the DIP already states that ambiguous ETI should be illegal. If you would like to propose that I change this, please tell me at least a rough idea of what you expect that to look like.
If you do not want this changed, then what you’re talking about is illegal syntax. I don’t really think we need to ponder which type of error the compiler will give us, I have no idea. If for some reason the implementation allows it, it will probably produce code that isn’t necessarily consistent between compilers, but that’s out of my hands as I am not implementing this feature—this is a DIP.

If you would like to discuss this matter further, please provide some insight into what you actually want out of this conversation, as I feel that it’s leading to nowhere. If you don’t actually want anything changed or revised, I can’t help you and I don’t really see the point in discussing this anymore.

Reply:
> The bottom line is that the DIP already states that ambiguous ETI should be illegal.

Yes, it's perfectly fine if the DIP says that the code we're discussing is ambiguous and should be rejected. As long as the DIP does not leave the implementer guessing whether the compiler should accept or reject a piece of code. You can give my code snippets as examples of code rejected as ambiguous in the "other considerations" section.